compiler/noirc_evaluator/src/brillig/brillig_gen/variable_liveness.rs

```diff
@@ -90,241 +86,273 @@ fn variables_used_in_block(block: &BasicBlock, dfg: &DataFlowGraph) -> Variables
+                    .expect("live_in for successor should have been calculated")
+                    .iter()
+                    .copied(),
+            );
         }
 
+        // Variables used in this block, defined in this block or before.
+        let used = variables_used_in_block(block, &func.dfg);
+
+        // Variables defined in this block are not alive at the beginning.
+        let defined = self.variables_defined_in_block(block_id, &func.dfg, constants);
+
+        // Live at the beginning are the variables used, but not defined in this block, plus the ones
+        // it passes through to its successors, which are used by them, but not defined here.
+        // (Variables used by successors and defined in this block are part of `live_out`, but not `live_in`).
+        let live_in = used.union(&live_out).filter(|v| !defined.contains(v)).copied().collect();
+
         // live_in[BlockId] = before_def[BlockId] union (live_out[BlockId] - killed[BlockId])
-        let passthrough_vars = live_out.difference(&defined).cloned().collect();
-        self.live_in.insert(block_id, used_before_def.union(&passthrough_vars).cloned().collect());
+        self.live_in.insert(block_id, live_in);
     }
 
-    fn compute_defined_variables(&self, block_id: BasicBlockId, dfg: &DataFlowGraph) -> Variables {
-        let block: &BasicBlock = &dfg[block_id];
+    /// Collects all the variables defined in a block, which includes:
+    /// * parameters of descendants this block immediately dominates
+    /// * the results of the instructions in the block
+    /// * constants which were allocated to this block
+    fn variables_defined_in_block(
+        &self,
+        block_id: BasicBlockId,
+        dfg: &DataFlowGraph,
+        constants: &ConstantAllocation,
+    ) -> Variables {
+        let block = &dfg[block_id];
         let mut defined_vars = HashSet::default();
 
-        for parameter in self.defined_block_params(&block_id) {
-            defined_vars.insert(parameter);
-        }
+        defined_vars.extend(self.defined_block_params(&block_id));
 
         for instruction_id in block.instructions() {
             let result_values = dfg.instruction_results(*instruction_id);
-            for result_value in result_values {
-                defined_vars.insert(*result_value);
-            }
+            defined_vars.extend(result_values.iter().copied());
         }
 
+        defined_vars.extend(constants.allocated_in_block(block_id));
+
         defined_vars
     }
 
-    fn update_live_ins_within_loop(&mut self, back_edge: BackEdge) {
-        let header_live_ins = self
-            .live_in
-            .get(&back_edge.header)
-            .expect("Live ins should have been calculated")
-            .clone();
+    /// Once we know which variables are alive before the block header,
+    /// we can visit all blocks contained in the loop and append those
+    /// as-is to all of them: since we know that we have to come back
```

Comment:
```suggestion
    /// we can append those variables to all of the loop's blocks.
    /// Since we know that we have to come back
```
I had to read the method again to discern between "those" and "them"